### PR TITLE
Add multi parallel verifiers to verify witness when new transaction received in Blockchain

### PR DESCRIPTION
**Pull request by eryeer**
According to Issue [#1188 Test result](https://github.com/neo-project/neo/issues/1188#issuecomment-547836572), when transaction is received in `Blockchain`, the process timespan is more than 1ms. That means the capablity of Blockchain to process new transaction is less than 1000 txs per second. 

Furthermore, the most time-consuming part of the transaction processing process is Witness verification(consume almost 1ms per tx). Since each transaction can perform its own witness verify, and does not interfere with other transactiosn, so we can parallelize Witness verification.

After pressure test, this PR's optimization can improve almost 200% TPS!

Process design is showed in UML.

微信截图_20191112171545

In this PR the parallel process is implemented by child Actors, called `ParallelVerifier`, created by `Blockchain` Actor. `Reverify` is executed before verify Witness to prevent the transaction with tx fee exceeding the account balance enters child actors for a meaningless Witness check. After Witness check passed, txs are sent back to `Blockchain` for other checks.

### CODE REVIEW

File: neo/Ledger/MemoryPool.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
+using System.Numerics;
```
Comment (shargon) [inline neo/Ledger/MemoryPool.cs]: This using is required?

File: neo/Ledger/VerifyingSenderFeeMonitor.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                _verifyingFeeRwLock.ExitReadLock();
+            }
+        }
+
+        public void AddSenderVerifyingFee(Transaction tx)
+        {
+            _verifyingFeeRwLock.EnterWriteLock();
+            try
+            {
+                if (!_senderVerifyingFee.ContainsKey(tx.Sender))
```
Comment (shargon) [inline neo/Ledger/VerifyingSenderFeeMonitor.cs]: Use `TryGet` in order to save one `get` in the `else`

File: neo/Ledger/MemoryPool.cs (hunk 1 above)
Comment (eryeer) [reply neo/Ledger/MemoryPool.cs]: No, I have deleted it.

**Comment by shargon** [comment]
We will test it soon, good work! 😊

**Comment by erikzhang** [comment]
This is too complicated.

Let's see the `Verify()` method of `Transaction`:

https://github.com/neo-project/neo/blob/927683d73ec7d244d4d3b7e633f2301150b8ccc4/neo/Network/P2P/Payloads/Transaction.cs#L210-L218

There are two steps in it:

1. Calling to the `Reverify()` method
2. The other parts of `Verify()` without calling `Reverify()`

Step 1 is lightweight and can be processed very quickly. So we can do it in `Blockchain`.

In step 2, we have to call `VerifyWitnesses()`, which needs to run NeoVM and is more resource intensive. So we can do it when the transactions are received in `ProtocolHandler`:

https://github.com/neo-project/neo/blob/927683d73ec7d244d4d3b7e633f2301150b8ccc4/neo/Network/P2P/ProtocolHandler.cs#L263-L267

In this way, the second step is parallel and the different nodes do not interact with each other.

**Comment by eryeer** [comment]
@erikzhang `ProtocolHandler` is a multi-instance Actor, but it can't take turns to verify transactions and share the problem of blocking the single thread's transaction verification.

If we put the Witness verification in the `ProtocolHandler`, then we can imagine a scenario: there are four nodes A, B, C, D. A large number of transactions are always initiated by node A, then the pressure of TPS will always exist in only one `ProtocolHandler` instance connected to A in each node, and other `ProtocolHandler` instances can not share the transaction verification work, then the verification will still block.

This problem does not occur if child Actor is used to take turns to receive transactions for verification.

**Comment by erikzhang** [comment]
In a real scenario, the transaction should come from multiple nodes. If a large number of transactions come from the same node, then the high probability is an attack. Using `ProtocolHandler` to handle transaction validation can reduce the impact of the attack to a certain extent.

**Comment by Tommo-L** [comment]
😂   Like the airdrop, they will create a log of transactions to one nodes, and the network topology also affects the delivery path, like following, most txs will be sent by path: A-->B --> CN1. 

image

And we should not assume that the nodes are in a completely ideal state.  The other is, do we have to put the logic processing on the protocol layer?

**Comment by erikzhang** [comment]
I think it is more beneficial to limit the airdrop to a frequency that does not affect the transactions sent by other nodes.

**Pull request closed without merging**